Modules/_fnmatchmodule.c

```diff
@@ -0,0 +1,246 @@
+/*
+ * C accelerator for the 'fnmatch' module.
+ *
+ * Most functions expect string or bytes instances, and thus the Python
+ * implementation should first pre-process path-like objects, and possibly
+ * applying normalizations depending on the platform if needed.
+ */
+
+#include "Python.h"
+
+#include "clinic/_fnmatchmodule.c.h"
+
+/*[clinic input]
+module _fnmatch
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=356e324d57d93f08]*/
+
+#include <fnmatch.h>
+
+static inline int
+validate_encoded_object(PyObject *name)
+{
+    if (!PyBytes_Check(name)) {
+        PyErr_Format(PyExc_TypeError,
+                     "name must be a bytes object, got %.200s",
+                     Py_TYPE(name)->tp_name);
+        return 0;
+    }
+    return 1;
+}
+
+static inline int
+validate_unicode_object(PyObject *name)
+{
+    if (!PyUnicode_Check(name)) {
+        PyErr_Format(PyExc_TypeError,
+                     "name must be a string object, got %.200s",
+                     Py_TYPE(name)->tp_name);
+        return 0;
+    }
+    return 1;
+}
+
+static inline int
+posix_fnmatch_encoded(const char *pattern, PyObject *name)
+{
+    if (!validate_encoded_object(name)) {
+        return -1;
+    }
+    // case-insensitive match
```

Comment:
Ok, this one was incorrect. It should have been case-sensititive (I messed up the correct function, my bad. I'll push a revision)